x/crosschain/keeper/gas_price.go

```diff
@@ -6,36 +6,72 @@ import (
 	"cosmossdk.io/math"
 	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"golang.org/x/exp/slices"
 
 	"github.com/zeta-chain/zetacore/x/crosschain/types"
 )
 
 // SetGasPrice set a specific gasPrice in the store from its index
 func (k Keeper) SetGasPrice(ctx sdk.Context, gasPrice types.GasPrice) {
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.GasPriceKey))
-	b := k.cdc.MustMarshal(&gasPrice)
 	gasPrice.Index = strconv.FormatInt(gasPrice.ChainId, 10)
+	b := k.cdc.MustMarshal(&gasPrice)
 	store.Set(types.KeyPrefix(gasPrice.Index), b)
 }
 
-// GetGasPrice returns a gasPrice from its index
-func (k Keeper) GetGasPrice(ctx sdk.Context, chainID int64) (val types.GasPrice, found bool) {
+// GetGasPrice returns a gasPrice from its index or false if it doesn't exist.
+func (k Keeper) GetGasPrice(ctx sdk.Context, chainID int64) (types.GasPrice, bool) {
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.GasPriceKey))
+
 	b := store.Get(types.KeyPrefix(strconv.FormatInt(chainID, 10)))
 	if b == nil {
-		return val, false
+		return types.GasPrice{}, false
 	}
+
+	var val types.GasPrice
 	k.cdc.MustUnmarshal(b, &val)
+
 	return val, true
 }
 
-func (k Keeper) GetMedianGasPriceInUint(ctx sdk.Context, chainID int64) (sdk.Uint, bool) {
-	gasPrice, isFound := k.GetGasPrice(ctx, chainID)
-	if !isFound {
-		return math.ZeroUint(), isFound
+// GetMedianGasPriceInUint returns the median gas price (and median priority fee) from the store
+// or false if it doesn't exist.
+func (k Keeper) GetMedianGasPriceInUint(ctx sdk.Context, chainID int64) (math.Uint, math.Uint, bool) {
+	entity, found := k.GetGasPrice(ctx, chainID)
+	if !found {
+		return math.ZeroUint(), math.ZeroUint(), false
+	}
+
+	var (
+		gasPrice    = math.NewUint(entity.Prices[entity.MedianIndex])
+		priorityFee = math.NewUint(medianValue(entity.PriorityFees))
+	)
+
+	return gasPrice, priorityFee, true
+}
+
+// medianValue returns the median value of a slice
+// example: [ 1 7 5 2 3 6 4 ] => [ 1 2 3 4 5 6 7 ] => 4
+func medianValue(items []uint64) uint64 {
+	switch len(items) {
+	case 0:
+		return 0
+	case 1:
+		return items[0]
 	}
-	mi := gasPrice.MedianIndex
-	return sdk.NewUint(gasPrice.Prices[mi]), true
+
+	// We don't want to modify the original slice
+	copiedItems := make([]uint64, len(items))
+	copy(copiedItems, items)
+
+	slices.Sort(copiedItems)
+	mv := copiedItems[len(copiedItems)/2]
```

Comment:
Given a data set of [1, 1, 1, 1, 7, 7, 7, 7] this would return 1 as the median value, is this expected? shouldn't we return 4 on that case?

Reply:
This is how it currently works for gasPrice 😅  Let's indeed use the "real" median there with some rounding to the closest integer. I'll also move this to a separate package.

Reply:
https://github.com/zeta-chain/node/pull/2483/commits/aab53c875f033855ac4ba568f0ec7b59fc2860f4